packages/mui-base/src/utils/useAnchorPositioning.ts

```diff
@@ -205,20 +204,29 @@ export function useAnchorPositioning(
     nodeId,
   });
 
-  // We can assume that element anchors are stable across renders, and thus can be reactive.
-  const reactiveAnchorDep = anchor == null || isElement(anchor);
-  const anchorRef = useLatestRef(anchor);
+  const rerender = useForcedRerendering();
+
+  const registeredPositionReference = React.useRef<Element | VirtualElement | null>(null);
 
   useEnhancedEffect(() => {
-    function isRef(param: unknown): param is React.MutableRefObject<any> {
-      return {}.hasOwnProperty.call(param, 'current');
+    const resolvedAnchor = typeof anchor === 'function' ? anchor() : anchor;
+
+    if (resolvedAnchor) {
+      const unwrappedElement = isRef(resolvedAnchor) ? resolvedAnchor.current : resolvedAnchor;
+      refs.setPositionReference(unwrappedElement);
+      registeredPositionReference.current = unwrappedElement;
     }
-    const resolvedAnchor =
-      typeof anchorRef.current === 'function' ? anchorRef.current() : anchorRef.current;
-    if (resolvedAnchor && !isElement(resolvedAnchor)) {
-      refs.setPositionReference(isRef(resolvedAnchor) ? resolvedAnchor.current : resolvedAnchor);
+  }, [refs, anchor]);
+
+  React.useEffect(() => {
+    // Refs from parent components are set after useLayoutEffect runs and are available in useEffect.
+    // Therefore, if the anchor is a ref, we need to update the position reference in useEffect.
+    const resolvedAnchor = typeof anchor === 'function' ? anchor() : anchor;
+    if (isRef(resolvedAnchor) && resolvedAnchor.current !== registeredPositionReference.current) {
+      refs.setPositionReference(resolvedAnchor.current);
```

Comment:
Is `flushSync` necessary now, if there will be a flicker?

Reply:
I haven't noticed it flickering. With the menu initially open, even with 20x CPU slowdown, the popup appears only near the anchor.